share/wake/lib/system/sources.wake

```diff
@@ -58,6 +58,67 @@ export def source (file: String): Result Path Error =
         x, Nil = Pass x
         _      = unreachable "a quoted RegExp can never match more than one distinct String"
 
+# Allows claiming of a file so long as the file is within
+# wakeroot. Keep in mind that if another job produces this
+# file, claim will fail. Likewise if you claim a file then
+# any job that produces this file will fail. Claim should
+# be thought of as a "virtual" job and wake demands that
+# all paths be produced by exactly one job, virtual or
+# otherwise.
+#
+# For source files please use `source`. For files outside
+# of wakeroot please use `claimFileAsPath` or `claimFileAsPathIn`.
+# `claim` is good for claiming files of previous builds that
+# are not being used again this build.
+#
+# Parameters:
+#  - `file`: The string giving the path to the file to be claimed
+#
+# Example:
+# ```
+#   claim "build/my_flow/my_step/previous-output.txt" # Claims the file
+# ```
+export def claim (file: String): Result Path Error =
+    # Define the actual virtual job that produces the Path
+    def raw_claim file =
+        def get_modtime file = prim "get_modtime"
+        def time = get_modtime file
+        if time == -1 then
+            failWithError "{file}: files does not exist. Maybe you haven't previously built this file?"
+        else
+            # Its important that the command here is distinct from what `source` would use
+            # so that claim and source do not overlap.
+            makeExecPlan ("<claim>", str time, file, Nil) Nil
+            | setPlanShare       False
+            | setPlanEcho        logDebug
+            | setPlanEnvironment Nil
+            | setPlanFnOutputs   (file, _)
+            | runJobWith virtualRunner
+            | getJobOutput
+
+    # Compute the relative path in simplist form
+    def isAbs = matches `/.*` file
+    def relPath = if isAbs then
+        relative workspace file
+    else
+        simplify file
+
+    # Fail if we're outside of wakeroot (relPath must be simplified)
+    def isOutsideRoot = matches `\.\..*/.*` relPath
+    require False = isOutsideRoot
+    else failWithError "claim cannot be used on files outside of wake's root. Consider using claimFileAsPath instead."
+
+    # Now that our path is uniform check if it could have been fetched with `source`
+    def base = basename relPath
+    def dir = dirname relPath
+    def scan dir regexp = prim "sources"
+    def canBeSourced = exists (_==~relPath) (scan dir base.quote)
+    require False = canBeSourced
+    else failWithError "{file}: Is a source file, please use `source` or `sources` instead of `claim`"
```

Comment:
Honestly, I'd love to remove the reliance on git.  My favorite VCS is Darcs, and as it is now, I wouldn't use Wake for personal projects.  I don't know exactly what went into designing `source`, but my guess at its merits would be not needing to walk the filesystem for basic file existence tests, ensuring that we can only `source` legal files (in the workspace, not a directory), and maybe something with filtering out `.gitignore` files.  Now that you've implemented those tests in Wake, I'm not sure we do need `source` any more, or rather that we could replace the behaviour of `source` with that of `claim`, though I'm certainly not the only one to ask.

Reply:
`export topic source: String` is an alternative to using `git`. You publish a list of strings of every file path that you would like to have be obtainable by the `sources` fn. I'm not suggesting we use that, but it exists if you want it for your Darcs usage you could create script that writes out a file containing what you want in your `publish source = ...`

Reply:
Ah, thanks!  (The other reason I'm not going to use Wake for personal projects is because Haskell already has its own build systems, but I'll keep that in mind if I ever try doing something in a different language.)